src/main/java/org/mockito/internal/invocation/InvocationMatcher.java

```diff
@@ -77,7 +77,7 @@ public String toString() {
 
     @Override
     public boolean matches(Invocation candidate) {
-        return invocation.getMock().equals(candidate.getMock())
+        return invocation.getMock() == candidate.getMock()
```

Comment:
I am not sure if I follow why this line was changed. Could you elaborate as to why?

Reply:
Consider the test PartialMockingWithSpiesTest.shouldAllowStubbingOfMethodsThatDelegateToOtherMethods.
Now the call to getName is handled by MockMethodInterceptor.DispatcherDefaultingToRealMethod.
interceptSuperCallable here the interceptor will not be null and it will go to interceptor.doIntercept. This finally leads to InvocationMatcher.matches

Now with changes in this pull request the call to equals is again delegated to  MockMethodInterceptor.DispatcherDefaultingToRealMethod and it will come at same place. And this way it leads to stackoverflow. Earlier equals of mock was this==that. So i just replced equals with ==.

Reply:
Awesome, thanks for the explanation! That makes sense to me 😄

Reply:
Haha Thanks. Had missed to put links to the classes.